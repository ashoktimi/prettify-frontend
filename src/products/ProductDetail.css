.ProductDetail{
    left: 0;
    right: 0;
    margin:auto;
    display: flex;
    justify-content: center;
}