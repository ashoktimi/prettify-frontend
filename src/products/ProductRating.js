import React from 'react';
import { faStar, faStarHalfAlt } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';


const ProductRating = ({ rating }) =>{

  const fullStars = Math.floor(rating);
  const partialStar = rating - fullStars;
  const emptyStars = 5 - fullStars - Math.ceil(partialStar);
  
  return (
      <>
        {Array.from({ length: fullStars }, (_, i) => (
            <FontAwesomeIcon
              key={i}
              icon={faStar}
              className="ProductDetail-star"
              style={{ color: '#fdd835' }}
            />
          ))}
          {partialStar > 0 && (
            <FontAwesomeIcon
              icon={faStarHalfAlt}
              className="ProductDetail-star"
              style={{ color: '#fdd835' }}
            />
          )}
          {Array.from({ length: emptyStars }, (_, i) => (
            <FontAwesomeIcon
              key={i + fullStars}
              icon={faStar}
              className="ProductDetail-star"
              style={{ color: 'lightgray' }}
            />
          ))}
     </>
    )  
}
  
export default ProductRating;