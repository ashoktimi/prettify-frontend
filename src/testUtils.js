import React from "react";
import UserContext from "./auth/userContext";


const demoUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
  photo_url: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, handleToggleOffcanvas = () => false }) => (
    <UserContext.Provider value={{ currentUser, handleToggleOffcanvas }}>
      {children}
    </UserContext.Provider>
);

function setup(){
  window.matchMedia = jest.fn().mockImplementation(query => {
    return {
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    };
  });
}

function teardown(){
   window.matchMedia.mockClear();
   delete window.matchMedia;
}

function commonBeforeEach() {
  setup();
}

function commonAfterEach() {
  teardown();
}

export { UserProvider, commonBeforeEach, commonAfterEach };