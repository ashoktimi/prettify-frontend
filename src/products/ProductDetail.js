import React, { useContext } from 'react';
import { useDispatch } from "react-redux";
import UserContext from "../auth/userContext";
import { addToCart,  } from "../cart/cartSlice";
import defaultImage from '../images/spinning image.png'
import ProductRating from './ProductRating';
import { Container, Row, Col, Image, Button } from 'react-bootstrap';

const ProductDetail = ({id, name, price, priceSign, prevPrice, description, image_link, rating, numOfRating, color }) => {
  const product = {
    id,
    name,
    price,
    priceSign,
    prevPrice,
    image_link,
    rating,
    numOfRating,
  };
  const { handleToggleOffcanvas } = useContext(UserContext);
  const dispatch = useDispatch();
  const handleAddToCart = (product) => {
    dispatch(addToCart(product));
    handleToggleOffcanvas();
  };
  const imageUrl = image_link || defaultImage;

  return (  
      <Container style={{marginTop: '5rem'}}>
        <Row className="justify-content-center align-items-center">
          <Col xs={12} md={6} className="text-center mb-4 mb-md-0">
            <Image
              className="ProductDetail-image"
              src={imageUrl}
              alt={name}
              onError={(e) => {
                e.target.src = defaultImage;
              }}
              style={{width: '30rem', height: '30rem', backgroundColor: 'white'}}
            />
          </Col>
          <Col xs={12} md={6}>
            <div className="ProductDetail-info">
              <h5 className="ProductDetail-name">{name}</h5>
              <p className="ProductDetail-description">{description}</p>
                <small className="ProductCard-price"> {priceSign}{price} </small> 
                <small className="ms-2 text-muted priceBefore"> before {priceSign}{prevPrice} </small>
                <div className="ProductDetail-rating">
                  <small>({rating})</small>
                  <ProductRating rating={rating} />
                  <p>({numOfRating}) Reviews</p>   
                  {color && (
                    <div>
                      {color.map(c => (
                        <div
                          key={c.id}
                          style={{
                            backgroundColor: `${c.hex_value}`,
                            width: '30px',
                            height: '30px',
                            borderRadius: '50%',
                            display: 'inline-block',
                            marginLeft: '5px'
                          }}
                          />
                      ))}
                    </div>
                  )}       
                </div>          
                <Button variant="warning"  onClick={() => handleAddToCart(product)} style={{marginTop:"2rem"}} >ADD To CART</Button>
            </div>
          </Col>
        </Row>
      </Container>
    )
  }

export default ProductDetail;