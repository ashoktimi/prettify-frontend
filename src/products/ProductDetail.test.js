import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { UserProvider,  commonBeforeEach, commonAfterEach } from "../testUtils";
import { configureStore } from '@reduxjs/toolkit';
import cartReducer from '../cart/cartSlice';
import ProductDetail from "./ProductDetail";
import ProductRating from './ProductRating';

beforeEach(commonBeforeEach)
afterEach(commonAfterEach)

describe('ProductCard', () => {
 const store = configureStore({
   reducer: {
     cart: cartReducer
   }
 });

 test("renders without crashing", function () {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <UserProvider>
            <ProductDetail />
          </UserProvider>          
        </Provider>    
      </MemoryRouter>   
     );
  });
  
  test("matches snapshot", function () {
    const { asFragment } = 
     render(
      <MemoryRouter>
        <Provider store={store}>
         <UserProvider>
           <ProductDetail />
         </UserProvider>          
        </Provider>    
      </MemoryRouter>  
     );
    expect(asFragment()).toMatchSnapshot();
  });
    
  test("ProductRating component renders correctly ", function () {
    render(
      <MemoryRouter>
         <ProductRating />
      </MemoryRouter>   
     );
  });
})
