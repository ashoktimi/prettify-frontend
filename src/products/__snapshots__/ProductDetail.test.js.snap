// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
    style="margin-top: 5rem;"
  >
    <div
      class="justify-content-center align-items-center row"
    >
      <div
        class="text-center mb-4 mb-md-0 col-md-6 col-12"
      >
        <img
          class="ProductDetail-image"
          src="spinning image.png"
          style="width: 30rem; height: 30rem; background-color: white;"
        />
      </div>
      <div
        class="col-md-6 col-12"
      >
        <div
          class="ProductDetail-info"
        >
          <h5
            class="ProductDetail-name"
          />
          <p
            class="ProductDetail-description"
          />
          <small
            class="ProductCard-price"
          >
              
          </small>
          <small
            class="ms-2 text-muted priceBefore"
          >
             before  
          </small>
          <div
            class="ProductDetail-rating"
          >
            <small>
              ()
            </small>
            <p>
              () Reviews
            </p>
          </div>
          <button
            class="btn btn-warning"
            style="margin-top: 2rem;"
            type="button"
          >
            ADD To CART
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
